source/agora/consensus/EnrollmentManager.d

```diff
@@ -1091,3 +1105,115 @@ unittest
+        };
+
+        auto signature = gen_key_pair.secret.sign(hashFull(tx)[]);
+        tx.inputs[0].signature = signature;
+        storage.put(tx);
+    }
+
+    // create an EnrollmentManager object
+    auto man = new EnrollmentManager(":memory:", key_pair);
+    scope (exit) man.shutdown();
+    Hash[] utxo_hashes = storage.keys;
+
+    Enrollment[] enrollments;
+    Enrollment enroll1;
+    Enrollment enroll2;
+    Enrollment enroll3;
+    ulong block_height = 2;
+
+    // create and add the first Enrollment object
+    auto utxo_hash1 = utxo_hashes[0];
+    assert(man.createEnrollment(utxo_hash1, block_height, enroll1));
+    assert(!man.hasEnrollment(utxo_hash1));
+    assert(man.add(enroll1, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
+    assert(man.hasEnrollment(utxo_hash1));
+    enrollments ~= enroll1;
+    assert(man.getValidatorCount(block_height) + enrollments.length == 1);
+
+    man.clearExpiredValidators(block_height);
+    foreach (ref enrollment; enrollments)
+        man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    enrollments.length = 0;
+    block_height = 3;
+
+    // create and add the second Enrollment object
+    auto utxo_hash2 = utxo_hashes[1];
+    assert(man.createEnrollment(utxo_hash2, block_height, enroll2));
+    assert(!man.hasEnrollment(utxo_hash2));
+    assert(man.add(enroll2, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
+    assert(man.hasEnrollment(utxo_hash2));
+    enrollments ~= enroll2;
+    assert(man.getValidatorCount(block_height) + enrollments.length == 2);
+
+    man.clearExpiredValidators(block_height);
+    foreach (ref enrollment; enrollments)
+        man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    enrollments.length = 0;
+    block_height = 4;
+
+    // create and add the third Enrollment object
+    auto utxo_hash3 = utxo_hashes[2];
+    assert(man.createEnrollment(utxo_hash3, block_height, enroll3));
+    assert(!man.hasEnrollment(utxo_hash3));
+    assert(man.add(enroll3, &storage.findUTXO));
+    assert(man.enroll_pool.count() == 1);
+    assert(man.hasEnrollment(utxo_hash3));
+    enrollments ~= enroll3;
+    assert(man.getValidatorCount(block_height) + enrollments.length == 3);
+
+    man.clearExpiredValidators(block_height);
+    foreach (ref enrollment; enrollments)
+        man.addValidator(enrollment, block_height, &storage.findUTXO);
+
+    enrollments.length = 0;
+    block_height = 5;    // valid block height : [0 .. 1008]
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 3);
+
+    block_height = 1009; // valid block height : [2 .. 1010]
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 3);
+
+    block_height = 1010; // valid block height : [3 .. 1011]
+    man.clearExpiredValidators(block_height);
+    assert(man.getValidatorCount(block_height) == 2);
```

Comment:
I think this should be 3 if because the first enrollment has the `enrolled_height` of 2 so the start height as a validator is 3 as you described. I have been thinking the `block_height` you passed in as parameter is the height where the new block is being made.
불필요한 번역시간을 줄이기 위해서 한글 코멘트를 남깁니다. 
이 assert 문에서 체크해야할 값은 3이 되어야한다고 생각합니다. 왜냐하면, 첫번째 enrollment의 enrolled_height는 2이고, validator로서 기능을 시작하는 것은 height가 3이 됩니다. 따라서, 첫번째 enrollment는 3부터 1010까지 validator로서 유효하기 때문에 위 assert 체크값은 3이 되어야 합니다.

Reply:
I thought after watching ClearExpiredValidators.
Please explain in detail what the problem is.
ClearExpiredValidators :
enrolled_height <= block_height - Enrollment.ValidatorCycle
getValidatorCount:
enrolled_height >= block_height - Enrollment.ValidatorCycle + 1